packages/tiktok-video-element/tiktok-video-element.js

```diff
@@ -0,0 +1,324 @@
+const EMBED_BASE = 'https://www.tiktok.com/player/v1';
+
+const serialize = (props) => {
+  const boolToBinary = (props) => {
+    let p = {};
+    for (let key in props) {
+      let val = props[key];
+      if (val === true || val === '') p[key] = 1;
+      else if (val === false) p[key] = 0;
+      else if (val != null) p[key] = val;
+    }
+    return p;
+  };
+  return String(new URLSearchParams(boolToBinary(props)));
+};
+
+function serializeIframeUrl(attrs, id, config = {}) {
+  const params = {
+    controls: attrs.controls === undefined ? null : 1,
+    autoplay: attrs.autoplay,
+    loop: attrs.loop,
+    rel: 0,
+    ...config,
+  };
+
+  return `${EMBED_BASE}/${id}?${serialize(params)}`;
+}
+
+function getTemplateHTML(attrs, id, config = {}) {
+  const src = serializeIframeUrl(attrs, id, config);
+  const dataConfigAttr = config ? `data-config='${JSON.stringify(config).replace(/'/g, '&apos;')}'` : '';
+
+  return /*html*/ `
+    <style>
+      :host { display:inline-block;position:relative;width:100%;height:100%; }
+      iframe { position:absolute;top:0;left:0;width:100%;height:100%;border:0;}
+    </style>
+    <iframe src="${src}" ${dataConfigAttr} allow="autoplay; fullscreen" title="TikTok video"></iframe>
+  `;
+}
+
+class TikTokVideoElement extends (globalThis.HTMLElement ?? class {}) {
+  static getTemplateHTML = getTemplateHTML;
+  static shadowRootOptions = { mode: 'open' };
+
+  #muted = false;
+  #trueMuted = false; // Used to track actual mute state
+
+  static get observedAttributes() {
+    return ['src', 'controls', 'loop', 'autoplay', 'muted'];
+  }
+
+  static PlayerState = { INIT: -1, ENDED: 0, PLAYING: 1, PAUSED: 2, BUFFERING: 3 };
+
+  constructor() {
+    super();
+    this.attachShadow({ mode: 'open' });
```

Comment:
here we should first check if there is already a shadow dom or not, in case it was SSR'd.
see other implementations